core/consensus/babe/src/lib.rs

```diff
@@ -75,28 +155,918 @@ impl Decode for BabeSeal {
+			}
+			CheckedHeader::Deferred(a, b) => {
+				debug!(target: "babe", "Checking {:?} failed; {:?}, {:?}.", hash, a, b);
+				telemetry!(CONSENSUS_DEBUG; "babe.header_too_far_in_future";
+					"hash" => ?hash, "a" => ?a, "b" => ?b
+				);
+				Err(format!("Header {:?} rejected: too far in the future", hash))
+			}
+		}
+	}
+}
+
+impl<B, C, E> Authorities<B> for BabeVerifier<C, E> where
+	B: Block,
+	C: ProvideRuntimeApi + ProvideCache<B>,
+	C::Api: AuthoritiesApi<B>,
+{
+	type Error = ConsensusError;
+
+	fn authorities(&self, at: &BlockId<B>) -> Result<Vec<AuthorityIdFor<B>>, Self::Error> {
+		authorities(self.client.as_ref(), at)
+	}
+}
+
+fn authorities<B, C>(client: &C, at: &BlockId<B>) -> Result<Vec<AuthorityIdFor<B>>, ConsensusError> where
+	B: Block,
+	C: ProvideRuntimeApi + ProvideCache<B>,
+	C::Api: AuthoritiesApi<B>,
+{
+	client
+		.cache()
+		.and_then(|cache| cache.get_at(&well_known_cache_keys::AUTHORITIES, at)
+			.and_then(|v| Decode::decode(&mut &v[..])))
+		.or_else(|| {
+			if client.runtime_api().has_api::<AuthoritiesApi<B>>(at).unwrap_or(false) {
+				AuthoritiesApi::authorities(&*client.runtime_api(), at).ok()
+			} else {
+				panic!("We don’t support deprecated code with new consensus algorithms, therefore this is unreachable; qed")
+			}
+		}).ok_or_else(|| consensus_common::ErrorKind::InvalidAuthoritiesSet.into())
+}
+
+/// The BABE import queue type.
+pub type BabeImportQueue<B> = BasicQueue<B>;
+
+/// Register the babe inherent data provider, if not registered already.
+fn register_babe_inherent_data_provider(
+	inherent_data_providers: &InherentDataProviders,
+	slot_duration: u64,
+) -> Result<(), consensus_common::Error> {
+	debug!(target: "babe", "Registering");
+	if !inherent_data_providers.has_provider(&srml_babe::INHERENT_IDENTIFIER) {
+		inherent_data_providers
+			.register_provider(srml_babe::InherentDataProvider::new(slot_duration))
+			.map_err(inherent_to_common_error)
+	} else {
+		Ok(())
+	}
+}
+
+fn get_keypair(q: &sr25519::Pair) -> &Keypair {
+	q.as_ref()
+}
+
+/// Pieces:
+///
+/// 1. VRF winnership
+/// 2. * VRF randomness
+///    * relative time
+
+fn make_transcript(
+	randomness: &[u8],
+	slot_number: u64,
+	genesis_hash: &[u8],
+	epoch: u64,
+) -> Transcript {
+	let mut transcript = Transcript::new(&BABE_ENGINE_ID);
+	transcript.proto_name(&BABE_ENGINE_ID);
+	transcript.commit_bytes(b"slot number", &slot_number.to_le_bytes());
+	transcript.commit_bytes(b"genesis block hash", genesis_hash);
+	transcript.commit_bytes(b"current epoch", &epoch.to_le_bytes());
+	transcript.commit_bytes(b"chain randomness", randomness);
+	transcript
+}
+
+/// Hash into chain:
+///
+/// * slot number
+/// * epoch
+/// * genesis hash
+/// * chain randomness
+fn author_block(
+	randomness: &[u8],
+	slot_number: u64,
+	genesis_hash: &[u8],
+	epoch: u64,
+	authorities: &[sr25519::Public],
+	key: &sr25519::Pair,
+	threshold: u64,
+) -> Option<(VRFInOut, VRFProof, VRFProofBatchable)> {
+	// FIXME this is O(n)
```

Comment:
1. I'm not convinced `O(n)` is actually a problem. It will take a good while to get over 1000, and then I think it's still fast to search.
2. all `FIXME` and `TODO` need links to filed substrate issues.

Reply:
I agree, and I filed the needed issues for FIXMEs.